### PR TITLE
Implement iadd,isub,imul for i128 in AArch64

### PR DESCRIPTION
**Pull request by afonso360**
This PR implements some basic operations (add, sub, mul) for i128 types on AArch64

### CODE REVIEW

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs
Comment (cfallin) [inline cranelift/codegen/src/isa/aarch64/lower_inst.rs:267]: I think we could get away with not using a `tmp` here, no? We could `umulh` directly into `dst_hi` and then use `madd` to add the other terms; that's slightly better from an allocate-fewer-vregs (compiler performance) perspective.

File: cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif
Comment (cfallin) [inline cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif:77]: A few more test vectors would be good to have here, I think -- e.g. the upper 64 bits of the first operand are always zero here. Want to make sure that all four of the half-product quadrants are computed properly :-)

File: cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif
Comment (cfallin) [inline cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif:14]: Could we include this test in a separate PR for the iconst issue? Or alternately just update the iconst implementation here (it should only be a few lines of code, I think!).

**Review by cfallin — COMMENTED** [review]
Thanks a bunch for this, @afonso360 ! 128-bit support on aarch64 has been on the TODO list for a while so I'm really happy to see someone start to chip away at this!

Just a few comments below but overall this looks good.

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs
Comment (akirilov-arm) [inline cranelift/codegen/src/isa/aarch64/lower_inst.rs:270]: This should be the second instruction in the sequence, since it can execute in parallel with `UMULH`.

File: cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif
Comment (afonso360) [reply cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif:14]: Yeah, I think I prefer including this in the PR for the iconst.

I think changing it here involves changing a bit more than just the emission for Iconst.  [`get_constant`](https://github.com/bytecodealliance/wasmtime/blob/76c6b83f6a21a12a11d4f890490f8acb6329a600/cranelift/codegen/src/machinst/lower.rs#L110) only returns a u64 and [`get_immediate`](https://github.com/bytecodealliance/wasmtime/blob/76c6b83f6a21a12a11d4f890490f8acb6329a600/cranelift/codegen/src/data_value.rs#L15) doesn't support 128bit values yet.

Comment (cfallin) [reply cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif:14]: Ah, yes, you're right, we need more infrastructure for that; sorry! Happy to have this whole test deferred then.

File: cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif
Comment (afonso360) [reply cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif]: I added a few more tests to all three functions, but if you have any particular test case in mind, let me know.

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs
Comment (afonso360) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:270]: It looks like when we do this, we do get a few extra mov's in the emitted code.

When placing the `mul` as the first or second instruction in the sequence we get the following code:
```
umulh x5, x0, x2
madd x4, x0, x2, xzr
madd x0, x0, x3, x5
madd x0, x1, x2, x0
mov x1, x0
mov x0, x4
```

Which makes sense, since we need x0 (lhs_lo) for the 2nd `madd`. I don't know if this is preferred to the 4 instruction codegen? It may be faster, but is longer?

Adding it after the second `madd` (as the third instruction in the sequence) gets us back to a 4 instruction result. But I'm not sure if `madd` executes in parallel with `mul`, I had a quick look at the Software Optimization guide for a few different cores and it doesn't look like it.

```
umulh x4, x0, x2
madd x3, x0, x3, x4
madd x0, x0, x2, xzr
madd x1, x1, x2, x3
```

Let me know which one you prefer

File: cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif
Comment (cfallin) [reply cranelift/filetests/filetests/isa/aarch64/arithmetic-run.clif]: the `C0FFEE` -> `DECAF` test vector is excellent; thanks :-)

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:270]: IMHO we should prefer code that keeps the regalloc happy and reduces moves (those can have a measurable impact as code-fetch bandwidth can be a limiting factor in some cases) to small reorderings, at least on modern out-of-order cores that can reason about dependencies and issue the `mul` as soon as possible. I'll defer to @akirilov-arm as to whether it makes sense to try harder here for older in-order microarchitectures though!

**Review by cfallin — APPROVED** [review]
Thanks! I'm happy with this code now, though let's finalize the question of instruction ordering before merging.

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs
Comment (akirilov-arm) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:270]: Are you testing with a simple function that just does the multiplication and returns the product? If yes, I wouldn't be too bothered by the additional moves because that's an overly simplified test case (and yes, the longer version might still be faster on an in-order core like the Cortex-A55 that I have in mind with my comments), but you can also try another experiment - reverse the order of the last 2 instructions and see what you get.

Comment (afonso360) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:270]: Yeah, these are all for the function that is in tests:
```
function %mul_i128(i128, i128) -> i128 {
block0(v0: i128, v1: i128):
    v2 = imul v0, v1
    return v2
}
```

> but you can also try another experiment - reverse the order of the last 2 instructions and see what you get.

Do you mean the last `mul` and `madd`? 

For the version with `mul` as the last instruction we get:
```
umulh x4, x0, x2
madd x3, x0, x3, x4
madd x1, x1, x2, x3
madd x0, x0, x2, xzr
```

For the version with the last `madd` as the last instruction, we get this:
```
umulh x4, x0, x2
madd x3, x0, x3, x4
madd x0, x0, x2, xzr
madd x1, x1, x2, x3
```

Comment (akirilov-arm) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:270]: Yes, I meant precisely your last experiment and the result is disappointing (the first move is unnecessary at the very least). Anyway, let's go with the current version, i.e. `MUL` in the last position (which still makes it possible to use the dedicated forwarding path between the pair of `MADD`s and to dual-issue), and I (or someone else) can check later what the real performance impact is and if there is some way to avoid the moves.

**Review by akirilov-arm — APPROVED** [review]

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs
Comment (akirilov-arm) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:270]: @cfallin Just a quick note - now that the announcement is [public](https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/first-armv9-cpu-cores), I can actually say that not only current generation, but also some of the next generation cores (i.e. Cortex-A510) are also going to be in-order.

**Pull request merged**
